### PR TITLE
Add sequence and numpy array indices support to `ndarray.take`

### LINKED ISSUE #6061: List and NumPy array support in chainerx.ndarray.take
Unlike CuPy, ChainerX's `ndarray.take()` doesn't accept `list` and `numpy.ndarray` as an `indices` argument yet. It is useful for implementing efficient dataset sampling feature in #5954.

```python
>>> import numpy as np
>>> import chainerx as chx
>>> a = chx.array(range(10), dtype=chx.float32)
>>> a.take(indices=chx.array([1, 3, 5]), axis=0)
array([1., 3., 5.], shape=(3,), dtype=float32, device='native:0')
>>> a.take(indices=[1, 3, 5], axis=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: take(): incompatible function arguments. The following argument types are supported:
    1. (self: chainerx.ndarray, indices: chainerx.ndarray, axis: Optional[int]=None) -> chainerx.ndarray

Invoked with: array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.], shape=(10,), dtype=float32, device='native:0'), [1, 3, 5]; kwargs: axis=0
>>> a.take(indices=np.array([1, 3, 5]), axis=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: take(): incompatible function arguments. The following argument types are supported:
    1. (self: chainerx.ndarray, indices: chainerx.ndarray, axis: Optional[int]=None) -> chainerx.ndarray

Invoked with: array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.], shape=(10,), dtype=float32, device='native:0'), array([1, 3, 5]); kwargs: axis=0
```

### PR DESCRIPTION
**Pull request by okapies**
- Fix to handle `sequence` and `numpy.ndarray` as indices in `ndarray.take`
    - This is the same behavior as NumPy and CuPy
    - It is useful for implementing batched dataset sampling algorithm (e.g. `a.take([...], axis=0)`)
- Add an overloaded `MakeArray` which receives C++ types

Note that the current implementation passes `list` to `MakeArray`, which converts `numpy.ndarray` by using `numpy.array` function and then converts it to chainerx array.

Fixes #6061.

## Benchmarks
- data: mnist train images (`(60000, 784)`)
- device: `native:0`
- number: `20`
- master branch: `010d676`

The benchmark script is [here](https://gist.github.com/okapies/fac7bf4a5971c092bccadb2face4284f).

| batch-size | chx (`master`) | chx | np | list | np (*1) | list (*1) |
|:--|:--|:--|:--|:--|:--|:--|
| 1 | 22.706 | 23.110 | 25.923 | 26.568 | 26.273 | 27.774 |
| 10 | 18.572 | 18.692 | 18.917 | 19.214 | 19.034 | 19.201 |
| 100 | 18.369 | 18.385 | 18.452 | 18.475 | 18.472 | 18.462 |
| 1000 | 18.038 | 18.174 | 18.166 | 18.124 | 18.182 | 18.255 |

(The values are rounded to three decimal places.)

*1: Convert a source index array to a chainerx array explicitly in Python

### CODE REVIEW

File: chainerx_cc/chainerx/python/array.h [hunk 1]
```diff
@@ -22,6 +24,8 @@ using ConstArrayBodyPtr = std::shared_ptr<const ArrayBody>;
 
 ArrayBodyPtr MakeArray(pybind11::handle object, pybind11::handle dtype, bool copy, pybind11::handle device);
 
+ArrayBodyPtr MakeArray(pybind11::handle object, nonstd::optional<Dtype>& dtype, bool copy, Device& device);
```
Comment (hvy) [inline chainerx_cc/chainerx/python/array.h]: Can you take the dtype by `const &` ?
```suggestion
ArrayBodyPtr MakeArray(pybind11::handle object, const nonstd::optional<Dtype>& dtype, bool copy, Device& device);
```

**Comment by hvy** [comment]
"Add tests" is under "TODO" in the descriptions but they seem to be covered?

**Comment by okapies** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by okapies** [comment]
@hvy Now all tests passed on both public CIs and Jenkins (without the spurious precision errors). Could you check?

**Comment by hvy** [comment]
LGTM!

**Comment by okapies** [comment]
@hvy Thanks!

**Pull request merged**